libs/wire-api-federation/src/Wire/API/Federation/Version.hs

```diff
@@ -66,6 +81,107 @@ instance ToSchema VersionInfo where
 versionInfo = VersionInfo (toList supportedVersions)
 
+----------------------------------------------------------------------
+
+data VersionRangeTag
+  = VersionRangeTagAll
+  | VersionRangeTagFrom
+  | VersionRangeTagUntil
+  | VersionRangeTagFromUntil
+  deriving (Eq, Enum, Bounded)
+
+versionRangeTagSchema :: ValueSchema NamedSwaggerDoc VersionRangeTag
+versionRangeTagSchema =
+  enum @Text "VersionRange Tag" $
+    mconcat
+      [ element "all-versions" VersionRangeTagAll,
+        element "from" VersionRangeTagFrom,
+        element "until" VersionRangeTagUntil,
+        element "from-until" VersionRangeTagFromUntil
+      ]
+
+versionPairSchema :: ValueSchema NamedSwaggerDoc (Version, Version)
+versionPairSchema =
+  object "VersionPair" $
+    (,)
+      <$> fst .= field "from" schema
+      <*> snd .= field "until" schema
+
+data VersionRange
+  = AllVersions
+  | FromVersion Version
+  | UntilVersion Version
+  | FromUntilVersion Version Version
+
+deriving instance Eq VersionRange
+
+deriving instance Show VersionRange
+
+deriving instance Ord VersionRange
+
+makePrisms ''VersionRange
+
+instance ToSchema VersionRange where
+  schema =
+    object "VersionRange" $
+      fromTagged
+        <$> toTagged
+          .= bind
+            (fst .= field "tag" versionRangeTagSchema)
+            (snd .= fieldOver _1 "value" untaggedSchema)
+    where
+      toTagged :: VersionRange -> (VersionRangeTag, VersionRange)
+      toTagged d@AllVersions = (VersionRangeTagAll, d)
+      toTagged d@(FromVersion _) = (VersionRangeTagFrom, d)
+      toTagged d@(UntilVersion _) = (VersionRangeTagUntil, d)
+      toTagged d@(FromUntilVersion _ _) = (VersionRangeTagFromUntil, d)
+
+      fromTagged :: (VersionRangeTag, VersionRange) -> VersionRange
+      fromTagged = snd
+
+      untaggedSchema = dispatch $ \case
+        VersionRangeTagAll -> tag _AllVersions null_
+        VersionRangeTagFrom -> tag _FromVersion (unnamed schema)
+        VersionRangeTagUntil -> tag _UntilVersion (unnamed schema)
+        VersionRangeTagFromUntil -> tag _FromUntilVersion $ unnamed versionPairSchema
+
+deriving via Schema VersionRange instance ToJSON VersionRange
+
+deriving via Schema VersionRange instance FromJSON VersionRange
+
+-- | Compute the lower and upper boundary of a version range. The first
+-- component of the pair is the lower boundary, while the second component is
+-- the upper boundary.
+versionRangeToBoundaries :: VersionRange -> (Version, Version)
+versionRangeToBoundaries AllVersions = (minBound @Version, maxBound @Version)
+versionRangeToBoundaries (FromVersion fv) = (fv, maxBound @Version)
+versionRangeToBoundaries (UntilVersion uv) = (minBound @Version, pred uv)
+versionRangeToBoundaries (FromUntilVersion fv uv) = (fv, pred uv)
+
+-- | Checks if a version is within a given version range.
+isWithinRange :: Version -> VersionRange -> Bool
+isWithinRange v vr =
+  let (lo, hi) = versionRangeToBoundaries vr
+   in lo <= v && v <= hi
```

Comment:
this means that either `from` or `until` (i keep forgetting which is inclusive) change their semantics to being inclusive.  not sure how i would resolve this.  probably go with the practice of the client api versioning combinators we all are already used to?

Reply:
Well spottted! I'm making the upper bound exclusive to make it aligned with the `Until` combinator.